src/SynapseWriter.php

```diff
@@ -0,0 +1,386 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Keboola\DbWriter\Synapse;
+
+use PDO;
+use SplFileInfo;
+use Keboola\DbWriter\Exception\ApplicationException;
+use Keboola\DbWriter\Exception\UserException;
+use Keboola\DbWriter\Synapse\Adapter\IAdapter;
+use Keboola\DbWriter\Writer;
+use Keboola\DbWriter\WriterInterface;
+use Psr\Log\LoggerInterface;
+
+class SynapseWriter extends Writer implements WriterInterface
+{
+    public const STAGE_NAME = '_db_writer_stage';
+
+    public const TEMP_NAME = '_db_writer_temp';
+
+    /** @var string[]  */
+    private static array $allowedTypes = [
+        'decimal', 'numeric', 'float', 'real', 'money', 'smallmoney', 'bigint', 'int', 'smallint',
+        'tinyint', 'bit', 'nvarchar', 'nchar', 'varchar', 'char', 'varbinary', 'binary', 'uniqueidentifier',
+        'datetimeoffset', 'datetime2', 'datetime', 'smalldatetime', 'date', 'time',
+    ];
+
+    /** @var string[]  */
+    private static array $typesWithSize = [
+        'decimal', 'numeric', 'float', 'nvarchar', 'nchar', 'varchar', 'char', 'varbinary', 'binary', 'time',
+    ];
+
+    protected IAdapter $adapter;
+
+    public static function getAllowedTypes(): array
+    {
+        return self::$allowedTypes;
+    }
+
+    public function __construct(array $dbParams, LoggerInterface $logger, IAdapter $adapter)
+    {
+        parent::__construct($dbParams, $logger);
+        $this->adapter = $adapter;
+    }
+
+    public function showTables(string $dbName): array
+    {
+        throw new ApplicationException('Method not implemented');
+    }
+
+    public function getTableInfo(string $tableName): array
+    {
+        throw new ApplicationException('Method not implemented');
+    }
+
+    public function write(SplFileInfo $csv, array $table): void
+    {
+        throw new ApplicationException('Method not implemented');
+    }
+
+    public function createConnection(array $dbParams): PDO
+    {
+        $host = $dbParams['host'] . ',' . $dbParams['port'];
+        $options['Server'] = 'tcp:' . $host;
+        $options['Database'] = $dbParams['database'];
+        $dsn = sprintf('sqlsrv:%s', implode(';', array_map(function ($key, $item) {
+            return sprintf('%s=%s', $key, $item);
+        }, array_keys($options), $options)));
+        $this->logger->info("Connecting to DSN '" . $dsn . "'");
+
+        $pdo = new PDO($dsn, $dbParams['user'], $dbParams['password'], [
+            'LoginTimeout' => 30,
+            'ConnectRetryCount' => 3,
+            'ConnectRetryInterval' => 10,
+            PDO::SQLSRV_ATTR_QUERY_TIMEOUT => 10800,
+        ]);
+        $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
+
+        return $pdo;
+    }
+
+    public function create(array $table): void
+    {
+        $with = [];
+        $primaryKey = $table['primaryKey'] ?? [];
+        if (count($primaryKey) === 1) {
+            $with[] = sprintf(
+                'DISTRIBUTION = HASH(%s)',
```

Comment:
Synapse nepodporuje `primary keys`, ale podporuje bezne kluce - indexy (roznym sposobom) + `HASH DISTRIBUTION`, ... to by malo ovplyvnovat hlavne `incremental load` aby bol dostatocne rychly, ... aby vedela DB rychlo najst zaznamy podla `primary key`, ktory je zadany v konfiguracii writera.

Bolo by to dobre overit s KBC, ci je to takto ok, ... oni maju realne skusenosti so Synapse. Su tam rozne obmedzenia, ... takze toto by som bral ako `v1`, ktoru treba zvalidovat.